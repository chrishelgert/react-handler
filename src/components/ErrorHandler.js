import React from 'react'
import PropTypes from 'prop-types'

/**
 * This Mini-Handler handles only the error-state
 * or display the given Component.
 */
const ErrorHandler = ({ message, ErrorComponent, children }) => {
  if (message) {
    if (ErrorComponent) {
      return <ErrorComponent>{message}</ErrorComponent>
    }

    return <div className='handler--error'>{message}</div>
  }

  return children
}

ErrorHandler.propTypes = {
  message: PropTypes.string,
  ErrorComponent: PropTypes.func,
  children: PropTypes.node
}

ErrorHandler.defaultProps = {
  message: null,
  ErrorComponent: null,
  children: null
}

export default ErrorHandler
