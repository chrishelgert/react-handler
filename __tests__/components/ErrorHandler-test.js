import React, { PropTypes } from 'react'
import renderer from 'react-test-renderer'
import ErrorHandler from '../../src/components/ErrorHandler'

const TestComponent = () => <h1>Test</h1>
const ErrorComponent = ({ children }) => <h2 className='error'>{children}</h2>
ErrorComponent.propTypes = { children: PropTypes.string }
const errorMessage = 'failed to call api'

describe('ErrorHandler', () => {
  test('returns the component when called without errorMessage', () => {
    const tree = renderer.create(
      <ErrorHandler>
        <TestComponent />
      </ErrorHandler>,
    )
    expect(tree.toJSON()).toMatchSnapshot()
  })

  test('returns null when called without errorMessage and component', () => {
    const tree = renderer.create(<ErrorHandler />)
    expect(tree).toMatchSnapshot()
  })

  test('returns the ErrorComponent', () => {
    const tree = renderer.create(
      <ErrorHandler
        ErrorComponent={ErrorComponent}
        message={errorMessage}
      >
        <TestComponent />
      </ErrorHandler>,
    )
    expect(tree.toJSON()).toMatchSnapshot()
  })

  test('returns the ErrorMessage when called without ErrorComponent', () => {
    const tree = renderer.create(
      <ErrorHandler message={errorMessage}>
        <TestComponent />
      </ErrorHandler>,
    )
    expect(tree.toJSON()).toMatchSnapshot()
  })
})
