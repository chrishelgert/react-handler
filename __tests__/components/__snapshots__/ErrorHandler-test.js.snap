// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorHandler should return null when called without errorMessage and component 1`] = `null`;

exports[`ErrorHandler should return the ErrorComponent 1`] = `
<h2
  className="error"
>
  failed to call api
</h2>
`;

exports[`ErrorHandler should return the ErrorMessage when called without ErrorComponent 1`] = `
<div
  className="handler--error"
>
  failed to call api
</div>
`;

exports[`ErrorHandler should return the component when called without errorMessage 1`] = `
<h1>
  Test
</h1>
`;
